subsys/bluetooth/host/att.c

```diff
@@ -189,11 +189,18 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf,
 
 	chan->sent = cb ? cb : chan_cb(buf);
 
+	/* bt_l2cap_send_cb takes onwership of the buffer so take another
+	 * reference to restore the state in case an error is returned.
```

Comment:
This isn't always true though, is it? If ownership was always taken you'd have an unconditional 'unref()', not a conditional one like you do now.

Reply:
Answering myself: yes, it's true. It's just that `bt_l2cap_send_cb()` and `chan_send()` have different buffer reference handling semantics, which is the cause of all this confusing code.